src/tpm/mod.rs

```diff
@@ -0,0 +1,158 @@
+mod tpm_wrapper;
+
+use std::{
+    convert::{TryFrom, TryInto},
+    sync::Once,
+};
+use std::sync::{Arc, Mutex};
+use thiserror::Error;
+use p256::{ecdsa};
+use p256::elliptic_curve::sec1::FromEncodedPoint;
+use sha2::{Digest, Sha256};
+
+use crate::{keypair, KeyTag, Network, public_key, Result, KeyType as CrateKeyType,
+            error, ecc_compact, ecc_compact::Signature};
+
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("function {0} returned error code {1}")]
+    TPMError(String, u32),
+
+    #[error("wrong key path")]
+    WrongKeyPath,
+}
+
+impl Error {
+    pub fn tpm_error(func: String, code: u32) -> crate::Error {
+        Self::TPMError(func, code).into()
+    }
+
+    pub fn wrong_key_path() -> crate::Error {
+        Self::WrongKeyPath.into()
+    }
+}
+
+
+static INIT: Once = Once::new();
+static mut COUNTER: Option<Arc<Mutex<i32>>> = None;
+
+pub struct Keypair {
+    pub network: Network,
+    pub public_key: public_key::PublicKey,
+    pub path: String,
+}
+
+impl PartialEq<Self> for Keypair {
+    fn eq(&self, other: &Self) -> bool {
+        self.public_key == other.public_key && self.path == other.path
+    }
+}
+
+impl std::fmt::Debug for Keypair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
+        f.debug_struct("Keypair")
+            .field("path", &self.path)
+            .field("public", &self.public_key)
+            .finish()
+    }
+}
+
+impl keypair::Sign for Keypair {
+    fn sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
+        use signature::Signer;
+        let signature = self.try_sign(msg)?;
+        Ok(signature.to_vec())
+    }
+}
+
+impl Drop for Keypair {
+    fn drop(&mut self) {
+        unsafe {
+            let mut num = COUNTER.as_ref().unwrap().lock().unwrap();
+            *num -= 1;
+            if *num == 0 {
+                tpm_wrapper::tpm_deinit();
+                COUNTER = None;
+            }
+        }
+    }
+}
+
+pub fn init() -> Result {
```

Comment:
Does this need to be `pub`? It doesn't look like the kinda of thing a consumer of this crate would call

Reply:
It is called from gateway-rs